### PR TITLE
Implements UserModel validation on onRegister method

### PR DESCRIPTION
**Pull request by bombozama**
This allows to completely control the validation message output. Specially useful for non english systems.

### CODE REVIEW

**Comment by bennothommo** [comment]
@bombozama Since the Model Validator trait already reads the `customMessages` and `attributeNames` properties of a model, you can simply create a plugin that extends the `User` model, and add a couple of dynamic properties into the model in order for the validator to display a custom message.

For example:
```php
UserModel::extend(function ($model) {
    $model->addDynamicProperty('customMessages', [
        // Your custom messages here
    ]);
    $model->addDynamicProperty('attributeNames', [
        // Your attribute names here
    ]);
});
```

**Comment by bombozama** [comment]
I added it because I tried to do it as you suggest and it does not grab neither the custom names or messages at all. Validation messages end up using the actual field name and the standard text in whatever language.

This is what I've attempted:
```php
public function boot() {
    \RainLab\User\Models\User::extend(function($model){

        // Extend the validation rules
        $model->rules['name'] = 'required|between:4,255';
        $model->rules['surname'] = 'required|between:3,255';

        // Extend the attribute names (just translated them to spanish)
        $model->addDynamicProperty('attributeNames', [
            'password' => 'contraseña',
            'name' => 'nombres',
            'surname' => 'apellidos'
        ]);

        // Extend the validation messages
        $model->addDynamicProperty('customMessages', [
            'password.required' => 'Custom required password message',
            'name.required' => 'Custom required name message',
            'surname.required' => 'Custom required surname message'
        ]);
    });
}
```

The result when posting is the following:
Screen Shot 2020-01-14 at 9 48 10 AM

File: components/Account.php [hunk 1]
```diff
@@ -287,12 +287,14 @@ public function onRegister()
             }
 
             $rules = (new UserModel)->rules;
+            $names = (new UserModel)->attributeNames;
+            $messages = (new UserModel)->customMessages;
```
Comment (LukeTowers) [inline components/Account.php]: Change above to:
```php
$model = new UserModel();
```

and then reference `$model->rules`, `$model->attributeNames`, `$model->customMessages` as required.

**Comment by mjauvin** [comment]
Yes, I suggested using the Model Validation here (https://github.com/rainlab/user-plugin/pull/414#issuecomment-567957656) .

I might actually implement this if you're willing to help testing it.

**Comment by bennothommo** [comment]
@bombozama Thanks for the further info. I think I would be more inclined to merge a fix that implements @mjauvin's [recommended changes](https://github.com/rainlab/user-plugin/pull/414#issuecomment-567957656). That random Validator probably should be dropped and replaced with the normal model validation so that it conforms to the same validation as other areas.

File: components/Account.php (hunk 1 above)
Comment (bombozama) [reply components/Account.php]: Changes implemented as suggested in 8a09dd279e01a98e72a5468d22c5e693e724947e

**Comment by LukeTowers** [comment]
@bombozama Could you implement @mjauvin's suggestion that @bennothommo linked to and test it out? That will go a long way to improving the consistency of how the user validation happens.

**Comment by bombozama** [comment]
Caught up with work. I'll push it in a few days.

**Comment by bombozama** [comment]
@mjauvin your solution on https://github.com/rainlab/user-plugin/pull/414#issuecomment-567957656 does the trick. 

I also removed the login settings check for email/username because it's already taken care of in the UserModel validation within the `beforeValidate` method (See https://github.com/rainlab/user-plugin/blob/master/models/User.php#L244)

Additionaly moved validation after IP address recording to allow extending validation to control that argument (could be useful on private systems, etc.).

Hope it helps.

**Comment by mjauvin** [comment]
The name of this PR should be changed since it doesn't reflect what it does anymore.

File: components/Account.php [hunk 2]
```diff
@@ -286,23 +286,19 @@ public function onRegister()
                 $data['password_confirmation'] = post('password');
             }
 
-            $rules = (new UserModel)->rules;
-
-            if ($this->loginAttribute() !== UserSettings::LOGIN_USERNAME) {
-                unset($rules['username']);
-            }
```
Comment (LukeTowers) [inline components/Account.php]: @bombozama this part is important to keep, you can change it to `unset($user->rules['username']);` and put it right before `$user->validate()`.

Comment (LukeTowers) [reply components/Account.php]: Hmm, took a look at the beforeValidate, seems kind of unnecessarily complicated. @bennothommo @daftspunk could you take a look at https://github.com/rainlab/user-plugin/commit/971ce8cdd20ab2b9c1bbad0b19af195bbcca30ac#r36870182 and give me your thoughts?

**Comment by mjauvin** [comment]
So, where are we on this?

File: components/Account.php (hunk 2 above)
Comment (mjauvin) [reply components/Account.php]: In my opinion, this shouldn't hold this PR to be processed. We should open a new PR for the issue you mention above.

Comment (bennothommo) [reply components/Account.php]: @mjauvin It does have a bearing on the functionality that @LukeTowers linked above, so we do have to pause the PR for it to be properly considered.

**Comment by daftspunk** [comment]
Validation messages are now pulled from the User model (8cf71759744b608bc4d78e8f774dda9e223e5455)

If you require validation of custom attributes, you should implement your own handler for this purpose. The attributes in this component are hard coded for security reasons where simplicity is required for an audited and controlled environment.

For example

```
function onRegister() {
    // Do validation of custom attributes here

    // Call component handler
    return $this->account->onRegister();
}
```

I hope this helps. Feel free to send us an email if you require further assistance.

**Pull request closed without merging**